modules/client/shared/src/main/scala/com/azavea/stac4s/api/client/SttpStacClientF.scala

```diff
@@ -1,42 +1,77 @@
 package com.azavea.stac4s.api.client
 
-import com.azavea.stac4s.{StacCollection, StacItem}
+import com.azavea.stac4s.{StacCollection, StacItem, StacLink, StacLinkType}
 
-import cats.MonadError
+import cats.MonadThrow
+import cats.syntax.apply._
+import cats.syntax.either._
 import cats.syntax.flatMap._
 import cats.syntax.functor._
 import cats.syntax.option._
 import eu.timepit.refined.types.string.NonEmptyString
+import fs2.Stream
 import io.circe.syntax._
-import io.circe.{Encoder, Json}
+import io.circe.{Encoder, Error, Json, JsonObject}
+import monocle.Lens
 import sttp.client3.circe.asJson
-import sttp.client3.{SttpBackend, basicRequest}
+import sttp.client3.{ResponseException, SttpBackend, UriContext, basicRequest}
 import sttp.model.Uri
 
-case class SttpStacClientF[F[_]: MonadError[*[_], Throwable], S: Encoder](
+case class SttpStacClientF[F[_]: MonadThrow, S: Lens[*, Option[PaginationToken]]: Encoder](
     client: SttpBackend[F, Any],
     baseUri: Uri
-) extends StacClientF[F, S] {
-  def search: F[List[StacItem]] = search(None)
+) extends StreamingStacClientF[F, Stream[F, *], S] {
+  private val paginationTokenLens = implicitly[Lens[S, Option[PaginationToken]]]
 
-  def search(filter: S): F[List[StacItem]] = search(filter.asJson.some)
+  /** Get the next page [[Uri]] from the retrieved [[Json]] body. */
+  private def getNextLink(body: Either[ResponseException[String, Error], Json]): F[Option[Uri]] =
+    body
+      .flatMap {
+        _.hcursor
+          .downField("links")
+          .as[Option[List[StacLink]]]
+          .map(_.flatMap(_.collectFirst { case l if l.rel == StacLinkType.Next => uri"${l.href}" }))
+      }
+      .liftTo[F]
 
-  private def search(filter: Option[Json]): F[List[StacItem]] =
-    client
-      .send {
-        filter
-          .fold(basicRequest)(f => basicRequest.body(f.asJson.noSpaces))
-          .post(baseUri.withPath("search"))
-          .response(asJson[Json])
+  def search: Stream[F, StacItem] = search(None)
+
+  def search(filter: S): Stream[F, StacItem] = search(filter.some)
+
+  private def search(filter: Option[S]): Stream[F, StacItem] = {
+    val emptyJson = JsonObject.empty.asJson
+    // the initial filter may contain the paginationToken that is used for the initial query
+    val initialBody = filter.map(_.asJson).getOrElse(emptyJson)
+    // the same filter would be used as a body for all pagination requests
+    val noPaginationBody = filter.map(paginationTokenLens.set(None)(_).asJson).getOrElse(emptyJson)
+    Stream
+      .unfoldLoopEval((baseUri.withPath("search"), initialBody)) { case (link, request) =>
+        client
+          .send(basicRequest.body(request.noSpaces).post(link).response(asJson[Json]))
+          .flatMap { response =>
+            val body  = response.body
+            val items = body.flatMap(_.hcursor.downField("features").as[List[StacItem]]).liftTo[F]
+            val next  = getNextLink(body).map(_.map(_ -> noPaginationBody))
```

Comment:
I guess this can be better handled via https://github.com/azavea/stac4s/issues/199 eventually